File: app/lib/models/ride.dart
import 'user.dart';

class Ride {
  final String id;
  final String ownerId;
  final String from;
  final String to;
  final DateTime timeStart;
  final DateTime timeEnd;
  final int seats;
  final String status;
  final DateTime createdAt;
  final User? owner;
  final int? pendingRequestCount;

  Ride({
    required this.id,
    required this.ownerId,
    required this.from,
    required this.to,
    required this.timeStart,
    required this.timeEnd,
    required this.seats,
    required this.status,
    required this.createdAt,
    this.owner,
    this.pendingRequestCount,
  });

  factory Ride.fromJson(Map<String, dynamic> json) {
    return Ride(
      id: json['_id'] ?? json['id'],
      ownerId: json['ownerId'] is Map
          ? json['ownerId']['_id']
          : json['ownerId'],
      from: json['from'],
      to: json['to'],
      timeStart: DateTime.parse(json['timeStart']),
      timeEnd: DateTime.parse(json['timeEnd']),
      seats: json['seats'] ?? 1,
      status: json['status'],
      createdAt: DateTime.parse(json['createdAt']),
      owner: json['owner'] != null ? User.fromJson(json['owner']) : null,
      pendingRequestCount: json['pendingRequestCount'],
    );
  }

  bool get isOpen => status == 'open';
  bool get isFilled => status == 'filled';
  bool get isCancelled => status == 'cancelled';
  bool get isExpired => DateTime.now().isAfter(timeEnd);
}

File: backend/src/models/Ride.js
const mongoose = require('mongoose');

const rideSchema = new mongoose.Schema({
    ownerId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    from: {
        type: String,
        required: [true, 'Starting point is required'],
        trim: true,
        maxlength: [100, 'Location cannot exceed 100 characters']
    },
    to: {
        type: String,
        required: [true, 'Destination is required'],
        trim: true,
        maxlength: [100, 'Location cannot exceed 100 characters']
    },
    timeStart: {
        type: Date,
        required: [true, 'Start time is required']
    },
    timeEnd: {
        type: Date,
        required: [true, 'End time is required']
    },
    seats: {
        type: Number,
        default: 1,
        min: [1, 'At least 1 seat required'],
        max: [4, 'Maximum 4 seats allowed']
    },
    status: {
        type: String,
        enum: ['open', 'filled', 'cancelled'],
        default: 'open'
    }
}, {
    timestamps: true
});

// Index for efficient querying of open rides
rideSchema.index({ status: 1, timeEnd: 1 });
rideSchema.index({ ownerId: 1 });

// Virtual to populate owner details
rideSchema.virtual('owner', {
    ref: 'User',
    localField: 'ownerId',
    foreignField: '_id',
    justOne: true
});

rideSchema.set('toJSON', { virtuals: true });
rideSchema.set('toObject', { virtuals: true });

module.exports = mongoose.model('Ride', rideSchema);
